src/main/java/org/tugraz/sysds/runtime/instructions/fed/AggregateBinaryFEDInstruction.java

```diff
@@ -81,95 +86,295 @@ public void processInstruction(ExecutionContext ec) {
 		// TODO compute matrix multiplication
 		// compute matrix-vector multiplication
 		AggregateBinaryOperator ab_op = (AggregateBinaryOperator) _optr;
-		if (mo1.isFederated())
-			federatedAggregateBinary(mo1, mo2, out, ab_op, false);
-		else if (mo2.isFederated())
-			federatedAggregateBinary(mo2, mo1, out, ab_op, true);
-		else
-			throw new DMLRuntimeException("Federated instruction on non federated data");
+		if (mo1.isFederated() && mo2.getNumColumns() == 1) {// MV
+			MatrixBlock vector = mo2.acquireRead();
+			federatedAggregateBinaryMV(mo1, vector, out, ab_op, true);
+			mo2.release();
+		}
+		else if (mo2.isFederated() && mo1.getNumRows() == 1) {// VM
+			MatrixBlock vector = mo1.acquireRead();
+			federatedAggregateBinaryMV(mo2, vector, out, ab_op, false);
+			mo1.release();
+		}
+		else // MM
+			federatedAggregateBinary(mo1, mo2, out);
 	}
 	
 	/**
 	 * Performs a federated binary aggregation (currently only MV and VM is supported).
-	 * 
-	 * @param mo1 the federated matrix object
-	 * @param mo2 the other matrix object (currently has to be a vector)
-	 * @param op the operation
-	 * @param swapParams if <code>this</code> and <code>other</code> should be swapped (other op this, instead of
-	 *                   this op other)
+	 *
+	 * @param mo1 the first matrix object
+	 * @param mo2 the other matrix object
+	 */
+	private void federatedAggregateBinary(MatrixObject mo1, MatrixObject mo2, MatrixObject out) {
```

Comment:
missing out in jdoc

Reply:
I don't quite understand what you mean by that.